### DISCUSSION TITLE
Stream remove listener

### DISCUSSION THREAD

**Discussion by develocode777** [serverpod/serverpod#2047] [Q&A] [answered]
Is it necessary to remove the listener set in `streamOpened` with `session.messages.addListener` when you close the connection? If yes how to do that correctly? Is it sufficient to specify the same channel name  or do you also need to pass the same function from addListener?

**Comment by develocode777** [comment]
So I looked into the code and it seems that you need to give the same function in `session.messages.removeListener` so I thought that I could save them with the `setUserObject` and then get them via `getUserObject` in `streamClosed` 

```
import 'package:serverpod/serverpod.dart';

class TestEndpoint extends Endpoint {
  @override
  bool get requireLogin => true;

  @override
  Future<void> streamOpened(StreamingSession session) async {
    String channel = 'channel';
    Listener listener = (SerializableEntity message) {
      sendStreamMessage(session, message);
    };
    session.messages.addListener(
      channel,
      listener,
    );
    UserObject userObject = UserObject(listeners: {channel: listener});
    setUserObject(session, userObject);
  }

  @override
  Future<void> streamClosed(StreamingSession session) async {
    UserObject userObject = getUserObject(session);

    userObject.listeners.forEach((channel, listener) {
      session.messages.removeListener(channel, listener);
    });
    setUserObject(session, null);
    return;
  }
}

class UserObject {
  final Map<String, Listener> listeners;

  UserObject({required this.listeners});
}

typedef Listener = void Function(SerializableEntity);
```

However I don't understand why this is not done by serverpod automatically when the stream is closed.

**Reply by SandPod** [reply, marked as answer]
We actually do remove the listeners when the session is closed. There is some more documentation and explanation around this in the Pixarama project, check out the comment here: https://github.com/serverpod/pixorama/blob/464a096d5fbe78e340a1f12b9db46314a1017a60/pixorama_server/lib/src/endpoints/pixorama_endpoint.dart#L127-L133

Some additional logic might be needed if the user should be able to subscribe and unsubscribe to multiple channels, we do that in our chat module for example: https://github.com/serverpod/serverpod/blob/1337e3aa8e93edd93aa51aa4e8bb88643b91d8b7/modules/serverpod_chat/serverpod_chat_server/lib/src/endpoints/chat_endpoint.dart#L50

The interface for our streams is not as user-friendly as it could be. We have a couple of suggestions for improving them that we will publish in an RFC soon making I bit easier to understand what is going on.

**Reply by develocode777** [reply]
Ah, that is great! Thank you.
